### ISSUE TITLE
training use MuSGD is almost twice as slow as  use SGD, is it normal?

### ISSUE THREAD

**Issue by zcf2020DPL** [ultralytics/ultralytics#23271]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

same model  and same data, only different optimizer, MuSGD vs SGD

**Comment by UltralyticsAssistant** [comment]
👋 Hello @zcf2020DPL, thank you for your interest in Ultralytics 🚀! This is an automated response to help you get started quickly, and an Ultralytics engineer will also assist soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it. In your case (MuSGD vs SGD speed), an MRE is especially helpful and should include:
-Your exact train command/code (Python or CLI) and all optimizer-related args
-Your `ultralytics` version, PyTorch version, CUDA info (if used), and OS
-Training logs showing the step time/epoch time for both runs
-Confirmation that everything else is identical (batch size, image size, AMP, workers, device, etc.)

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Laughing-q** [comment]
@zcf2020DPL yeah it's slower than normal SGD since in Muon it needs to calculate quintic Newton-Schulz iteration for some parts of the model parameters
https://github.com/ultralytics/ultralytics/blob/f2d3aed634a5b0e4828024718d4a61ab2f83fb19/ultralytics/optim/muon.py#L94-L94

**Comment by 00f8t** [comment]
> [@zcf2020DPL](https://github.com/zcf2020DPL) yeah it's slower than normal SGD since in Muon it needs to calculate quintic Newton-Schulz iteration for some parts of the model parameters
> 
> [ultralytics/ultralytics/optim/muon.py](https://github.com/ultralytics/ultralytics/blob/f2d3aed634a5b0e4828024718d4a61ab2f83fb19/ultralytics/optim/muon.py#L94-L94)
> 
> Line 94 in [f2d3aed](/ultralytics/ultralytics/commit/f2d3aed634a5b0e4828024718d4a61ab2f83fb19)
> 
>  update = zeropower_via_newtonschulz5(update)

what scenarios its worth using it then?

**Comment by glenn-jocher** [comment]
Muon-based updates add extra matrix ops, so MuSGD will usually be slower per-iteration, but it’s worth using when you care more about stability and “time-to-metric” than raw step throughput (e.g., longer schedules, larger models, harder datasets/augs where SGD is more likely to oscillate/diverge); in Ultralytics it’s also only applied to `param.ndim >= 2` tensors while the rest still get standard SGD-style updates (see the [MuSGD/Muon implementation reference](https://docs.ultralytics.com/reference/optim/muon/) and the [trainer optimizer parameter-grouping logic](https://docs.ultralytics.com/reference/engine/trainer/)). If you’re doing short runs or your training is clearly throughput-limited and SGD is already stable, sticking with `optimizer=SGD` is totally reasonable—just compare both under the same wall-clock budget rather than the same epoch count.

**Comment by saltedsword** [comment]
> Muon-based updates add extra matrix ops, so MuSGD will usually be slower per-iteration, but it’s worth using when you care more about stability and “time-to-metric” than raw step throughput (e.g., longer schedules, larger models, harder datasets/augs where SGD is more likely to oscillate/diverge); in Ultralytics it’s also only applied to `param.ndim >= 2` tensors while the rest still get standard SGD-style updates (see the [MuSGD/Muon implementation reference](https://docs.ultralytics.com/reference/optim/muon/) and the [trainer optimizer parameter-grouping logic](https://docs.ultralytics.com/reference/engine/trainer/)). If you’re doing short runs or your training is clearly throughput-limited and SGD is already stable, sticking with `optimizer=SGD` is totally reasonable—just compare both under the same wall-clock budget rather than the same epoch count.

Thank you for the detailed and patient explanation. Based on my experiments, using the same dataset and identical training settings, SGD reached a mAP@0.5:0.95 of 0.424 at epoch 284, while MuSGD surpassed 0.425 already at epoch 128. In terms of wall-clock time, 128 epochs with MuSGD took approximately 19.07 hours, which is clearly lower than the 23.19 hours required for 284 epochs with SGD. This result is quite encouraging😄.

Given MuSGD’s faster and more stable convergence behavior, I’m considering adjusting the early stopping patience. Previously, I set patience to 100; now I’m thinking of reducing it to 50, with the goal of achieving comparable or even slightly better performance without exceeding the previous total training time. For our current engineering workflow, training time is very sensitive, and achieving the absolute best possible metric is not strictly required.

Do you think this adjustment is reasonable in your experience, or would you recommend a different strategy to better balance final accuracy and overall training time?

**Comment by glenn-jocher** [comment]
Yes—dropping `patience` from 100 to ~50 is a reasonable way to trade a bit of peak metric for more predictable wall-clock, since early stopping in Ultralytics is simply “stop after N epochs with no fitness improvement” (see the [`EarlyStopping` reference](https://docs.ultralytics.com/reference/utils/torch_utils/) and the [`train` args](https://docs.ultralytics.com/modes/train/)); I’d just keep `patience` comfortably larger than any “late training phase change” you rely on (e.g. `close_mosaic`) so you don’t stop right before the usual end-of-run cleanup gains.

```bash
yolo train model=yolo26n.pt data=your.yaml optimizer=MuSGD epochs=400 patience=50 time=23.0
```

**Comment by saltedsword** [comment]
> Yes—dropping `patience` from 100 to ~50 is a reasonable way to trade a bit of peak metric for more predictable wall-clock, since early stopping in Ultralytics is simply “stop after N epochs with no fitness improvement” (see the [`EarlyStopping` reference](https://docs.ultralytics.com/reference/utils/torch_utils/) and the [`train` args](https://docs.ultralytics.com/modes/train/)); I’d just keep `patience` comfortably larger than any “late training phase change” you rely on (e.g. `close_mosaic`) so you don’t stop right before the usual end-of-run cleanup gains.
> 
> yolo train model=yolo26n.pt data=your.yaml optimizer=MuSGD epochs=400 patience=50 time=23.0

Thanks for the suggestion. Using a hard wall-clock limit with time is a very practical solution, and it aligns well with our time-sensitive engineering constraints.

**Comment by glenn-jocher** [comment]
Makes sense—`time=` is a good “hard cap” (it’s in hours) and you can keep `epochs` high while letting `patience` decide when to stop early; just note `time` can stop you mid-epoch, so if you’re doing tight wall-clock comparisons it can help to set a periodic checkpoint via `save_period` and rely on `best.pt` for the final pick; the rest of the useful knobs are in the [training configuration guide](https://docs.ultralytics.com/usage/cfg/).

**Comment by APeFerreira** [comment]
I've also noticed that the time per epoch during the initial epochs is indeed around the double (21 min) of what used to be with SGD (11 min), for the dataset I'm training (+100k images, YOLO size: medium). However, I've noticed that, with MuSGD, the time per epoch grows with epochs (e.g., at epoch 50, each epoch takes 40 min). This didn't happened with SGD, since all epochs took around 11 min.
@saltedsword have you also noticed this behavior?
@glenn-jocher is this expected to happen?

**Comment by glenn-jocher** [comment]
Not expected—MuSGD adds mostly constant extra math per step (see the [MuSGD/Muon implementation reference](https://docs.ultralytics.com/reference/optim/muon/)), so epoch time shouldn’t steadily climb; that pattern usually points to validation time being included in your “per-epoch” number, a data pipeline/cache/paging slowdown, or an environment mismatch. Can you paste your exact `yolo train ...` command plus a text log snippet from ~epoch 1 and ~epoch 50, and also run:

```bash
yolo checks
```

**Comment by APeFerreira** [comment]
> Not expected—MuSGD adds mostly constant extra math per step (see the [MuSGD/Muon implementation reference](https://docs.ultralytics.com/reference/optim/muon/)), so epoch time shouldn’t steadily climb; that pattern usually points to validation time being included in your “per-epoch” number, a data pipeline/cache/paging slowdown, or an environment mismatch. Can you paste your exact `yolo train ...` command plus a text log snippet from ~epoch 1 and ~epoch 50, and also run:
> 
> ```
> yolo checks
> ```

I have made some minor modifications to the Ultralytics source code, mainly to the _fitness_ score, such as printing it and parametrizing the `fit_weights`. I haven't changed any code related to training, which is the troublesome part.

I have started my training with the arguments below
```yaml
mode: "train"

task: "detect"
batch: 52
epochs: 500
patience: 50
imgsz: 640

degrees: 7.5
translate: 0.1
mosaic: 0.0
fliplr: 0.0
flipud: 0.0
cutmix: 0.0

scale: 0.25
mask_ratio: 1
overlap_mask: false
deterministic: false

seed: 42
resume: true
save_period: -1
name: "."
visualize: false
save: true
plots: true
cache: false
```

Here's an excerpt of the first training epoch
```
Starting training for 500 epochs...

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss total_loss  Instances       Size
      1/500        35G     0.7626        5.7   0.008017       6.47         23        640           : 100% ━━━━━━━━━━━━ 1832/1832 1.4it/s 22:22
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)    Fitness: 100% ━━━━━━━━━━━━ 227/227 2.4it/s 1:36
                   all      23567      23568      0.813      0.705      0.807      0.644      0.762
```

And of epoch 64
```
      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss total_loss  Instances       Size
     64/500      35.1G     0.3708     0.2389   0.003737     0.6135         23        640           : 100% ━━━━━━━━━━━━ 1832/1832 1.3s/it 39:45
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)    Fitness: 100% ━━━━━━━━━━━━ 227/227 1.1s/it 3:58
                   all      23567      23568      0.946       0.96      0.989      0.924      0.961
```

I got this by running `yolo checks`
```
Ultralytics 8.4.8 🚀 Python-3.11.14 torch-2.9.1+cu128 CUDA:0 (NVIDIA A100-SXM4-40GB, 40326MiB)
Setup complete ✅ (6 CPUs, 83.5 GB RAM, 279.5/491.9 GB disk)

OS                     Linux-5.10.0-38-cloud-amd64-x86_64-with-glibc2.35
Environment            Docker
Python                 3.11.14
Install                git
Path                   /ultralytics/ultralytics
RAM                    83.48 GB
Disk                   279.5/491.9 GB
CPU                    Intel Xeon CPU @ 2.20GHz
CPU count              6
GPU                    NVIDIA A100-SXM4-40GB, 40326MiB
GPU count              1
CUDA                   12.8

numpy                  ✅ 2.4.2>=1.23.0
matplotlib             ✅ 3.10.8>=3.3.0
opencv-python-headless ✅ 4.13.0.90>=4.6.0
pillow                 ✅ 12.1.0>=7.1.2
pyyaml                 ✅ 6.0.3>=5.3.1
requests               ✅ 2.32.5>=2.23.0
scipy                  ✅ 1.17.0>=1.4.1
torch                  ✅ 2.9.1+cu128<2.10,>=1.8.0
torch                  ✅ 2.9.1+cu128!=2.4.0,<2.10,>=1.8.0; sys_platform == "win32"
torchvision            ✅ 0.24.1+cu128>=0.9.0
psutil                 ✅ 7.2.2>=5.8.0
polars                 ✅ 1.37.1>=0.20.0
ultralytics-thop       ✅ 2.0.18>=2.0.1
```

**PS:** I don't know if it makes any difference, but I have interrupted my training at epoch 17 and then resumed it with `last.pt`. This trend of increasing time per epoch started after resuming.

**Comment by glenn-jocher** [comment]
That steady jump from ~0.71 s/it (epoch 1) to ~1.3 s/it (epoch 64) isn’t expected from MuSGD (its extra Newton–Schulz work is essentially constant per step), and the fact that **val also slows down** points more to system/I/O/clock throttling or a code-path sync than the optimizer itself; since it started after an interrupt/resume, could you paste the log snippet from just before you stopped (e.g. epoch 16–17) and just after resuming (epoch 18–19), plus the exact resume command you ran? If you want to quickly isolate where time is growing, you can also add a small timing probe with Ultralytics’ [`Profile` utility](https://docs.ultralytics.com/reference/utils/ops/) around dataloading vs forward/backward/`optimizer_step()` and share the timings.

```bash
yolo train model=path/to/last.pt data=your.yaml optimizer=MuSGD resume=True
```

**Comment by APeFerreira** [comment]
@glenn-jocher thanks for the help! But finally I think the cause of the issue was interrupting and then resuming the training. I've started a new training from scratch, and epoch time has remained constant overall. I just can't get a similar performance with YOLOv26 compared to YOLOv11, but that's another type of problem I guess 😅

**Comment by glenn-jocher** [comment]
Good to hear starting fresh kept epoch time stable; if you can reproduce the slowdown specifically with `resume=True` (i.e. same run, same machine), please share the exact stop/resume commands and the epoch-17/18 log pair and we can dig in. For the YOLO26 vs YOLO11 accuracy gap (separate topic), please open a new issue for visibility and include your exact `yolo train ...` commands for both, `yolo checks`, and the run artifacts (`args.yaml` + `results.csv` from each run) so we can compare recipes apples-to-apples; the knobs most likely to explain differences are all in the [training configuration guide](https://docs.ultralytics.com/usage/cfg/).
